service/worker/sysworkflow/history_blob_iterator.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package sysworkflow
+
+import (
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+)
+
+/**
+IMPORTANT: Under the assumption that history is immutable, the following iterator will deterministically return identical
+blobs from Next regardless of the current cluster and regardless of the number of times this iteration occurs. This property
+makes any concurrent uploads of history safe.
+*/
+
+type (
+	// HistoryBlobIterator is used to get history blobs
+	HistoryBlobIterator interface {
+		Next() (*HistoryBlob, error)
+		HasNext() bool
+	}
+
+	historyBlobIterator struct {
+		logger        bark.Logger
+		metricsClient metrics.Client
+
+		// the following defines the state of the iterator
+		blobPageToken        int
+		persistencePageToken []byte
+		finishedIteration    bool
+
+		// the following are only used to read history and dynamic config
+		historyManager       persistence.HistoryManager
+		historyV2Manager     persistence.HistoryV2Manager
+		domainID             string
+		workflowID           string
+		runID                string
+		eventStoreVersion    int32
+		branchToken          []byte
+		lastFirstEventID     int64
+		config               *Config
+		domain               string
+		clusterName          string
+		closeFailoverVersion int64
+	}
+)
+
+// NewHistoryBlobIterator returns a new HistoryBlobIterator
+func NewHistoryBlobIterator(
+	logger bark.Logger,
+	metricsClient metrics.Client,
+	request ArchiveRequest,
+	container *SysWorkerContainer,
+	domainName string,
+	clusterName string,
+) HistoryBlobIterator {
+	return &historyBlobIterator{
+		logger:        logger,
```

Comment:
Can you make sure the logger is tagged with useful tags like domain, workflowId, runId, etc

Reply:
Yes it is, the logger we use here is the same logger that we tag in the activity which has these useful tags defined.